sources/core/Stride.Core/Diagnostics/Profiler.cs

```diff
@@ -56,37 +58,119 @@ namespace Stride.Core.Diagnostics
     public static class Profiler
     {
-        private class RollingBuffer
+        internal class ProfilingEventChannel
         {
-            private const int RollingBufferCount = 2;
-            private readonly List<FastList<ProfilingEvent>> rollingBuffers = new List<FastList<ProfilingEvent>>(RollingBufferCount);
-            private int currentBufferIndex;
+            internal static ProfilingEventChannel Create(UnboundedChannelOptions options)
+            {
+                var channel = Channel.CreateUnbounded<ProfilingEvent>(options);
+
+                return new ProfilingEventChannel { _channel = channel };
+            }
+
+            private Channel<ProfilingEvent> _channel;
+
+            internal ChannelWriter<ProfilingEvent> Writer => _channel.Writer;
+            internal ChannelReader<ProfilingEvent> Reader => _channel.Reader;
+        }
+
+        private class ThreadEventCollection
+        {
+            //TODO (perf): Use an SPSC queue instead.
+            private ProfilingEventChannel channel = ProfilingEventChannel.Create(new UnboundedChannelOptions { SingleReader = true, SingleWriter = true });
 
-            public RollingBuffer()
+            internal ThreadEventCollection()
             {
-                for (int i = 0; i < RollingBufferCount; i++)
-                    rollingBuffers.Add(new FastList<ProfilingEvent>());
+                Profiler.AddThread(this);
             }
 
-            public FastList<ProfilingEvent> GetBuffer()
+            internal void Add(ProfilingEvent e)
             {
-                return rollingBuffers[currentBufferIndex];
+                if (subscriberChannels.Count > 0)
+                    channel.Writer.TryWrite(e);
             }
 
-            public void MoveToNextBuffer()
+            internal IAsyncEnumerable<ProfilingEvent> ReadEvents()
             {
-                currentBufferIndex = (currentBufferIndex + 1) % RollingBufferCount;
-                rollingBuffers[currentBufferIndex].Clear();
+                return channel.Reader.ReadAllAsync();
             }
         }
 
+        //TODO: Hack. No guaranteees this won't collide with a CPU thread id.
+        //      Fix when taking another look at GPU profiling.
+        internal const int GpuThreadId = -1;
+
         internal static Logger Logger = GlobalLogger.GetLogger("Profiler"); // Global logger for all profiling
-        
-        private static readonly RollingBuffer CpuEvents = new RollingBuffer();
-        private static readonly RollingBuffer GpuEvents = new RollingBuffer();
+        internal static TimeSpan StartTime = TimeSpanFromTimeStamp(Stopwatch.GetTimestamp());
+        internal static TimeSpan GpuStartTime = TimeSpan.Zero;
+
         private static readonly object Locker = new object();
         private static bool enableAll;
         private static int profileId;
+        private static ThreadLocal<ThreadEventCollection> events = new(() => new ThreadEventCollection(), true);
```

Comment:
We use the ThreadLocal to let us use a more efficient SingleWriter, SingleReader Channel for the initial event submission and push the synchronisation to a different task. In practice the underlying SWSR Channel is still MWSR with a lock for writing, so we could still go lower level to an SPSC queue, but the runtime doesn't provide one and it would make the implementation more complex.

The ThreadId is set to the ManagedThreadId when creating the ProfilingScope and this would only be incorrect if you later manually move the ProfilingScope to a different thread. Profiling async code across `await` will give incorrect results and would need a different solution.

Reply:
Ok, so it is a threading requirement. Cool, I think the way you went about it should work.